src/core/Config.cpp

```diff
@@ -130,109 +411,65 @@ Config::Config(const QString& fileName, QObject* parent)
         init(portablePath);
-    } else {
-        QString userPath;
-        QString homePath = QDir::homePath();
-
-#if defined(Q_OS_UNIX) && !defined(Q_OS_MACOS)
-        // we can't use QStandardPaths on X11 as it uses XDG_DATA_HOME instead of XDG_CONFIG_HOME
-        QByteArray env = qgetenv("XDG_CONFIG_HOME");
-        if (env.isEmpty()) {
-            userPath = homePath;
-            userPath += "/.config";
-        } else if (env[0] == '/') {
-            userPath = QFile::decodeName(env);
-        } else {
-            userPath = homePath;
-            userPath += '/';
-            userPath += QFile::decodeName(env);
-        }
+        return;
+    }
 
-        userPath += "/keepassxc/";
+    QString configPath;
+    QString localConfigPath;
+
+#if defined(Q_OS_WIN)
+    configPath = QDir::fromNativeSeparators(QStandardPaths::writableLocation(QStandardPaths::AppDataLocation));
+    localConfigPath =
+        QDir::fromNativeSeparators(QStandardPaths::writableLocation(QStandardPaths::AppLocalDataLocation));
+#elif defined(Q_OS_MACOS)
+    configPath = QDir::fromNativeSeparators(QStandardPaths::writableLocation(QStandardPaths::AppDataLocation));
+    localConfigPath = QStandardPaths::writableLocation(QStandardPaths::CacheLocation);
 #else
-        userPath = QDir::fromNativeSeparators(QStandardPaths::writableLocation(QStandardPaths::DataLocation));
-        // storageLocation() appends the application name ("/keepassxc") to the end
-        userPath += "/";
+    configPath = QDir::fromNativeSeparators(QStandardPaths::writableLocation(QStandardPaths::AppConfigLocation));
+    localConfigPath = QStandardPaths::writableLocation(QStandardPaths::CacheLocation);
 #endif
 
+    configPath += "/keepassxc";
+    localConfigPath += "/keepassxc";
+
 #ifdef QT_DEBUG
-        userPath += "keepassxc_debug.ini";
-#else
-        userPath += "keepassxc.ini";
+    configPath += "_debug";
+    localConfigPath += "_debug";
 #endif
 
-        init(userPath);
-    }
+    configPath += ".ini";
+    localConfigPath += ".ini";
+
+    init(QDir::toNativeSeparators(configPath), QDir::toNativeSeparators(localConfigPath));
 }
 
 Config::~Config()
 {
 }
 
-void Config::init(const QString& fileName)
+void Config::init(const QString& configFileName, const QString& localConfigFileName)
 {
-    m_settings.reset(new QSettings(fileName, QSettings::IniFormat));
-    upgrade();
-    connect(qApp, &QCoreApplication::aboutToQuit, this, &Config::sync);
+    // Upgrade from previous KeePassXC version which stores its config
+    // in AppData/Local on Windows instead of AppData/Roaming.
+    // Move file to correct location before continuing.
+    if (!localConfigFileName.isEmpty() && QFile::exists(localConfigFileName) && !QFile::exists(configFileName)) {
+        QDir().mkpath(QFileInfo(configFileName).absolutePath());
+        QFile::copy(localConfigFileName, configFileName);
+        QFile::remove(localConfigFileName);
+        QDir().rmdir(localConfigFileName);
```

Comment:
This will not remove the directory, you need to get the absolute dir path of the filename and pass it to this function. QDir functions are very stupid in this regard.